### PR TITLE
librbd: make flush be queued by QOS throttler

### PR DESCRIPTION
**Pull request by trociny**
Previously we started tracking it only since it was enqueued and
the flush operation was skipping those aio, which had visible
effect if QOS was enabled.

Fixes: https://tracker.ceph.com/issues/38706
Signed-off-by: Mykola Golub <mgolub@suse.com>

- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

### CODE REVIEW

**Comment by trociny** [comment]
@dillaman It still requires mock tests cleanup and just has been preliminary tested. But I would like to know if my approach is what you had in your mind. 

I am not sure about this `finish_op->start_op` hack that I  had to add when an io detects the lock acquire or image refresh is needed (because the refresh at the end calls flush which would dead lock).

File: src/librbd/io/ImageDispatchSpec.h [hunk 1]
```diff
@@ -128,6 +128,7 @@ class ImageDispatchSpec {
   bool is_write_op() const;
 
   void start_op();
+  void finish_op();
```
Comment (dillaman) [inline src/librbd/io/ImageDispatchSpec.h]: Perhaps a different name? The op isn't being finished, it's being deferred or paused.

**Comment by trociny** [comment]
@dillaman Testing revealed yet another issue with starting async op so early. If `ImageRequestWQ::block_writes` comes at the moment when an aio is queued and not enqueued yet, `block_writes` will proceed with `flush_image` (because `m_in_flight_writes` is not increased for that aio yet) and gets stuck. it is sporadically reproduced with `TestInternal.ShrinkFlushesCache` and I believe the cause of the current Jenkins failure.

I tired to move `m_in_flight_writes++` to `start_in_flight_io` (so the `block_writes` wold just register its context in the `write_blocker` queue and returns) but it did not help much: that racing aio can not be enqueued and processed due to `m_write_blockers > 0`. It seems that I would need to track if the request was added to the queue (`start_in_flight_io`) when io was not blocked yet and then check this flag in `_void_deque` and allow the request to proceed even if `m_write_blockers > 0`. If we want to go this way...

**Comment by dillaman** [comment]
@trociny Maybe the better way to handle this is to tweak the throttle to allow "0" cost items to still be queued if there are already other items stuck in the queue. Then ```ImageDispatchSpec::TokenRequestedVisitor``` / ```ImageDispatchSpec::tokens_requested``` could return the cost via an out param and return true or false if the item should be queued. The ```Flush``` spec would always return a cost of zero but return ```true``` if QoS is enabled.

**Comment by trociny** [comment]
@dillaman I can try your suggestion. But I am just thinking about a different approach. Wouldn't we want to move the throttler to a dedicated dispatcher layer (similarly to what I am doing with simple io scheduler)?

**Comment by dillaman** [comment]
@trociny That would work, but right now we don't have an ```ImageDispatcher``` (planned but not yet implemented). The ```ImageRequestWQ``` is doing all the awkward heavy lifting for acquiring the lock and refreshing the image. You wouldn't want to schedule it at the ```ObjectDispatcher``` level because you would potentially double-count large ops.

**Comment by trociny** [comment]
Ah, indeed, we would need `ImageRequestDispatcher`, not `ObjectDispatcher` for this.

**Comment by trociny** [comment]
@dillaman updated. I am still testing it but it seems to work. Thank you for your suggestion!

**Comment by trociny** [comment]
Actually when I try it on not fixed rbd (which does not wait for ios complete before flush if interrupted) like this:
```
rbd bench --io-type write --io-size 4K --io-total 10M test --rbd-cache=false --rbd_qos_iops_limit 100 --debug-rbd=30
```
it still crashes. Interestingly in the log I see that the `handle_throttle_ready` for the flush request is called the last as expected, but by some reason it is processed earlier than a write requested issued before it:
```
   -45> 2019-03-14 19:44:01.019 7fafbaffd700 15 librbd::io::ImageRequestWQ: 0x55d131d0b720 handle_throttle_ready: r=0, req=0x55d131f4fe90
   -44> 2019-03-14 19:44:01.019 7fafbaffd700 15 librbd::io::ImageRequestWQ: 0x55d131d0b720 handle_throttle_ready: r=0, req=0x55d131f54ea0
   -43> 2019-03-14 19:44:01.019 7fafbb7fe700 20 librbd::io::ImageRequestWQ: 0x55d131d0b720 process: ictx=0x55d131f0d900, req=0x55d131f54ea0
   -42> 2019-03-14 19:44:01.019 7fafbb7fe700 20 librbd::io::ImageRequest: 0x7fafbb7f8fa0 send: aio_flush: ictx=0x55d131f0d900, completion=0x55d131f519e0
   -41> 2019-03-14 19:44:01.019 7fafbb7fe700 20 librbd::io::AioCompletion: 0x55d131f519e0 set_request_count: pending=1
   -40> 2019-03-14 19:44:01.019 7fafbb7fe700 20 librbd::io::ObjectDispatcher: 0x55d131f55730 send: object_dispatch_spec=0x7fafb001e700
   -39> 2019-03-14 19:44:01.019 7fafbb7fe700 20 librbd::io::AsyncOperation: 0x55d131f51b20 start_op
   -38> 2019-03-14 19:44:01.019 7fafbb7fe700 20 librbd::io::AioCompletion: 0x55d131f519e0 complete_request: cb=1, pending=0
   -37> 2019-03-14 19:44:01.019 7fafbb7fe700 20 librbd::io::AioCompletion: 0x55d131f519e0 finalize: r=0
   -36> 2019-03-14 19:44:01.019 7fafbb7fe700 20 librbd::io::AsyncOperation: 0x55d131f51b20 finish_op
   -35> 2019-03-14 19:44:01.019 7fafbb7fe700 20 librbd::io::AsyncOperation: 0x55d131e09600 start_op
   -34> 2019-03-14 19:44:01.019 7fafbb7fe700 20 librbd::io::ImageRequestWQ: 0x55d131d0b720 process: ictx=0x55d131f0d900, req=0x55d131f4fe90
   -33> 2019-03-14 19:44:01.019 7fafbb7fe700 20 librbd::io::ImageRequest: 0x7fafbb7f8f30 send: aio_write: ictx=0x55d131f0d900, completion=0x55d131e094c0
```
note, `req=0x55d131f4fe90` is `aio_write` and `req=0x55d131f54ea0` is `aio_flush` in this log.

**Comment by dillaman** [comment]
@trociny aargh ... it's because ```PointerWQ::requeue``` just does a push-front since it didn't used to need to worry about multiple ops in-flight.

**Comment by dillaman** [comment]
... simple solution might just be to kick the delayed IO requests to the back of the queue upon ```handle_throttle_ready```. Worst case it might need to process a hundred or so IO ops before it gets back to the original one (but it should have queued up all write-ops when the first one got queued so the actual number should be much lower).

**Comment by trociny** [comment]
@dillaman After preliminary testing it seems to work. Don't we want to add a special requeue method to `ThreadPool::PointerWQ` for this particular case? Probably `requeue_back`? And then may be rename the existing `requeue` method to `requeue_front` to make it clear what it does?

**Comment by dillaman** [comment]
> Don't we want to add a special requeue method to `ThreadPool::PointerWQ` for this particular case? Probably `requeue_back`? And then may be rename the existing `requeue` method to `requeue_front` to make it clear what it does?

Sounds good to me.

**Comment by trociny** [comment]
@dillaman testing unittest_librbd with qos enabled, something like this:
```
CEPH_ARGS='--rbd_qos_iops_limit=10' ./bin/unittest_librbd --gtest_filter=TestInternal.ShrinkFlushesCache
```
revealed an issue with flushing in resize request: in `ResizeRequest::send_flush_cache` we bypass the image request work queue and call the dispatcher directly. As a result this flush request ignores requests in the throttler queue. I am going to try modifying `ImageRequestWQ<I>::aio_flush` to accept additional `FlushSource flush_source` parameter, and use it in the `ResizeRequest::send_flush_cache` instead. Or do you have some other idea?

This also reminds me about a thing I was forgetting to ask for some time: with `rbd_non_blocking_aio` option set to `false` we skip the queue too and the QOS throttling does not work at all. Do you think it is ok? I am not sure how `rbd_non_blocking_aio=false` is used, may be it is already deprecated and we may not bother with this? But we could for example ignore `non_blocking_aio` setting when throttling is used and always use the queue.

**Comment by dillaman** [comment]
> in `ResizeRequest::send_flush_cache` we bypass the image request work queue and call the dispatcher directly. As a result this flush request ignores requests in the throttler queue

Seems like an odd edge condition to get into, right? It implies that you are shrinking an image while you have active IO beyond the extents of the newly shrunk image? What happens? Does the IO just fail w/ ```-EINVAL``` since it fails the bounds check?

> This also reminds me about a thing I was forgetting to ask for some time: with `rbd_non_blocking_aio` option set to `false` we skip the queue too and the QOS throttling does not work at all. Do you think it is ok? 

Yeah, I wouldn't worry about it, honestly. I am pretty sure nobody uses it and I was asked to create that option as a fallback after introducing the work queue. I would personally like to see that option go away in the Octopus release (semi-related to the non-blocking throttler changes I am making in Objecter).

**Comment by trociny** [comment]
> Seems like an odd edge condition to get into, right? It implies that you are shrinking an image while you have active IO beyond the extents of the newly shrunk image? What happens? Does the IO just fail w/ `-EINVAL` since it fails the bounds check?

Probably I was not very clear. So the failed test:
```
% RBD_FETATURES=127 CEPH_ARGS='--debug-rbd=30 --rbd_qos_iops_limit=5' ./bin/unittest_librbd --gtest_filter=TestInternal.ShrinkFlushesCache
seed 28964
Note: Google Test filter = TestInternal.ShrinkFlushesCache
[==========] Running 1 test from 1 test case.
[----------] Global test environment set-up.
[----------] 1 test from TestInternal
[ RUN      ] TestInternal.ShrinkFlushesCache
/home/mgolub/ceph/ceph/src/test/librbd/test_internal.cc:821: Failure
Value of: c->is_complete()
  Actual: false
Expected: true
```
The test sends aio_write just before resizing the image and then check after resize that aio was complete (the aio was flushed before resize). And it fails, which does not look odd to me, taking that `ResizeRequest::send_flush_cache` calls `io::ImageDispatchSpec<I>::create_flush_request` directly, that is bypassing the image worker (and QOS throttler) queue.

The problem is not introduced by my change, it is reproduced in master.  I just pushed a commit to this branch to show how I propose to fix this (the patch is not complete -- mock tests need tweaking).

**Comment by dillaman** [comment]
To me that just seems like a bad test expectation. I think the intent was just to verify that in-flight IOs were flushed, not that any submitted IOs were flushed. When we shrink, we don't want an in-flight IO to (re)create an object that is now outside the bounds of the shrink image.

**Comment by trociny** [comment]
Ah I see now what you mean. In this case yes it looks like the test just is wrong. 

I removed that patch and repushed.

**Pull request merged**